File: engine/crates/ppcap-core/src/columnar/schema.rs
//! The canonical flow Parquet/Arrow schema — the single source of truth.
//!
//! **29 columns.** Column order here == on-disk Parquet order == the DuckDB `flow` view's
//! SELECT order. The `schema_drift` test (CI guard) asserts all three agree. Any column
//! change MUST bump [`FLOW_PARQUET_VERSION`] and update the SQL view + `flow_columns_in_order`.
//!
//! This module is FULLY IMPLEMENTED (it is contract, not algorithm).

use std::sync::Arc;

use arrow_schema::{DataType, Field, Schema, TimeUnit};

/// On-disk flow schema version, written to the Parquet footer KV metadata
/// (`ppcap.flow_schema_version`) for forward-compatibility and external tooling. In Phase 0
/// the only flow reader is external DuckDB (`read_parquet` in `sql/schema.sql`), which does
/// not inspect this value; there is no in-engine reader that enforces it. Bump it whenever a
/// column is added/removed/reordered.
pub const FLOW_PARQUET_VERSION: u16 = 8;

/// Canonical Arrow schema for the persisted flow Parquet table.
pub fn flow_arrow_schema() -> Arc<Schema> {
    let utc = || DataType::Timestamp(TimeUnit::Nanosecond, Some("UTC".into()));
    Arc::new(Schema::new(vec![
        Field::new("flow_id", DataType::UInt64, false), // 1  monotonic id (assigned at write)
        Field::new("capture_id", DataType::UInt64, false), // 2
        Field::new("src_ip", DataType::Utf8, false), // 3  initiator (= lo endpoint), canonical string
        Field::new("dst_ip", DataType::Utf8, false), // 4  responder (= hi endpoint)
        Field::new("src_port", DataType::UInt16, false), // 5  lo_port; 0 for portless L4
        Field::new("dst_port", DataType::UInt16, false), // 6  hi_port; 0 for portless L4
        Field::new("proto", DataType::UInt8, false), // 7  IANA L4 proto (6/17/1/58/132...)
        Field::new("app_proto", DataType::Utf8, true), // 8  "dns"/"https"/...; NULL if unknown
        Field::new("bytes_c2s", DataType::UInt64, false), // 9  == bytes_fwd (lo->hi)
        Field::new("bytes_s2c", DataType::UInt64, false), // 10 == bytes_rev (hi->lo)
        Field::new("pkts", DataType::UInt64, false), // 11 pkts_fwd + pkts_rev
        Field::new("start_ts", utc(), false),        // 12 first_ts_ns (UTC ns)
        Field::new("end_ts", utc(), false),          // 13 last_ts_ns  (UTC ns)
        Field::new("tcp_flags_c2s", DataType::UInt8, false), // 14 tcp_flags_fwd
        Field::new("tcp_flags_s2c", DataType::UInt8, false), // 15 tcp_flags_rev
        Field::new("ttl_min_c2s", DataType::UInt8, false), // 16 ttl_min_fwd
        Field::new("category", DataType::Utf8, false), // 17 snake_case token; never NULL ("unknown")
        Field::new("app_proto_src", DataType::Utf8, true), // 18 derivation: "port"/"payload"; NULL when neither
        Field::new("sni", DataType::Utf8, true),           // 19 TLS SNI host; NULL if none observed
        Field::new("ja3", DataType::Utf8, true), // 20 TLS JA3 fingerprint; NULL if none observed
        Field::new("ja4", DataType::Utf8, true), // 21 TLS JA4 fingerprint; NULL if none observed
        Field::new("tls_version", DataType::Utf8, true), // 22 negotiated TLS version label; NULL if none
        Field::new("tls_cipher", DataType::Utf8, true), // 23 negotiated cipher-suite label; NULL if none
        Field::new("hassh", DataType::Utf8, true), // 24 SSH client HASSH (MD5) fingerprint; NULL if none
        Field::new("hassh_server", DataType::Utf8, true), // 25 SSH server HASSHServer (MD5); NULL if none
        Field::new("ja3s", DataType::Utf8, true), // 26 TLS JA3S server fingerprint (MD5); NULL if none
        Field::new("severity", DataType::Utf8, false),  // 27 lowercase token, never NULL ("info")
        Field::new("threat_score", DataType::UInt16, false), // 28 0..=100
        Field::new("ioc", DataType::Boolean, false),    // 29 any feed match on this flow
    ]))
}

/// CI drift guard: exact column names in canonical order.
pub fn flow_columns_in_order() -> [&'static str; 29] {
    [
        "flow_id",
        "capture_id",
        "src_ip",
        "dst_ip",
        "src_port",
        "dst_port",
        "proto",
        "app_proto",
        "bytes_c2s",
        "bytes_s2c",
        "pkts",
        "start_ts",
        "end_ts",
        "tcp_flags_c2s",
        "tcp_flags_s2c",
        "ttl_min_c2s",
        "category",
        "app_proto_src",
        "sni",
        "ja3",
        "ja4",
        "tls_version",
        "tls_cipher",
        "hassh",
        "hassh_server",
        "ja3s",
        "severity",
        "threat_score",
        "ioc",
    ]
}

File: ui/src/lib/query/schema.ts
/**
 * Browser-side flow schema for the NLQ query engine — the single UI source for
 * DuckDB column types, nullability, and per-column semantics (the semantics feed
 * the NL→SQL prompt in a later phase).
 *
 * Column NAMES and ORDER come from {@link FLOW_COLUMNS} (canonical Parquet order,
 * engine schema v10). This module only attaches DuckDB types to them; the
 * `Record<FlowColumn, …>` key type makes the compiler reject a missing or extra
 * column, and schema.test.ts guards order against the shared flow_columns.json
 * fixture (which the engine's `schema_drift` test also checks).
 */

import { FLOW_COLUMNS, type FlowColumn } from "../../types";

/**
 * Mirrors `FLOW_PARQUET_VERSION` in engine/crates/ppcap-core/src/columnar/schema.rs.
 * Bump in lockstep whenever the engine bumps (the fixture test enforces agreement).
 */
export const FLOW_SCHEMA_VERSION = 10;

/** DuckDB column types used by the browser `flow` table. */
export type DuckDbType =
  | "UBIGINT"
  | "USMALLINT"
  | "UTINYINT"
  | "VARCHAR"
  | "TIMESTAMP"
  | "BOOLEAN";

export interface FlowColumnSpec {
  type: DuckDbType;
  nullable: boolean;
  /** One-line semantics, phrased for the NL→SQL system prompt. */
  comment: string;
}

/**
 * Types/semantics per column. The engine's Parquet stores `start_ts`/`end_ts` as
 * ns-precision TIMESTAMPTZ; the browser table is built from normalized FlowRow
 * milliseconds, so plain TIMESTAMP (UTC wall clock, ms precision) is truthful here.
 */
export const FLOW_COLUMN_TYPES: Record<FlowColumn, FlowColumnSpec> = {
  flow_id: { type: "UBIGINT", nullable: false, comment: "monotonic flow id, unique per capture" },
  capture_id: { type: "UBIGINT", nullable: false, comment: "capture this flow belongs to" },
  src_ip: { type: "VARCHAR", nullable: false, comment: "initiator IP (SYN sender / first packet)" },
  dst_ip: { type: "VARCHAR", nullable: false, comment: "responder IP" },
  src_port: { type: "USMALLINT", nullable: false, comment: "initiator port; 0 for portless L4" },
  dst_port: { type: "USMALLINT", nullable: false, comment: "responder port; 0 for portless L4" },
  proto: { type: "UTINYINT", nullable: false, comment: "IANA L4 protocol number (6=TCP, 17=UDP, 1=ICMP, 58=ICMPv6, 132=SCTP)" },
  app_proto: { type: "VARCHAR", nullable: true, comment: "application protocol token, e.g. 'dns'/'https'; NULL if unknown" },
  bytes_c2s: { type: "UBIGINT", nullable: false, comment: "bytes initiator→responder (uploads)" },
  bytes_s2c: { type: "UBIGINT", nullable: false, comment: "bytes responder→initiator (downloads)" },
  pkts: { type: "UBIGINT", nullable: false, comment: "total packets, both directions" },
  start_ts: { type: "TIMESTAMP", nullable: false, comment: "first packet time, UTC" },
  end_ts: { type: "TIMESTAMP", nullable: false, comment: "last packet time, UTC" },
  tcp_flags_c2s: { type: "UTINYINT", nullable: false, comment: "OR of initiator-side TCP flags" },
  tcp_flags_s2c: { type: "UTINYINT", nullable: false, comment: "OR of responder-side TCP flags" },
  ttl_min_c2s: { type: "UTINYINT", nullable: false, comment: "minimum initiator-side IP TTL" },
  category: { type: "VARCHAR", nullable: false, comment: "traffic category token (see category list); never NULL ('unknown')" },
  app_proto_src: { type: "VARCHAR", nullable: true, comment: "how app_proto was derived: 'payload' or 'port'; NULL when neither" },
  sni: { type: "VARCHAR", nullable: true, comment: "TLS SNI host from the ClientHello" },
  ja3: { type: "VARCHAR", nullable: true, comment: "TLS JA3 client fingerprint (MD5)" },
  ja4: { type: "VARCHAR", nullable: true, comment: "TLS JA4 client fingerprint" },
  tls_version: { type: "VARCHAR", nullable: true, comment: "negotiated TLS version label" },
  tls_cipher: { type: "VARCHAR", nullable: true, comment: "negotiated TLS cipher-suite label" },
  hassh: { type: "VARCHAR", nullable: true, comment: "SSH client HASSH fingerprint (MD5)" },
  hassh_server: { type: "VARCHAR", nullable: true, comment: "SSH server HASSHServer fingerprint (MD5)" },
  ja3s: { type: "VARCHAR", nullable: true, comment: "TLS JA3S server fingerprint (MD5)" },
  http_host: { type: "VARCHAR", nullable: true, comment: "HTTP request Host header" },
  http_ua: { type: "VARCHAR", nullable: true, comment: "HTTP request User-Agent header" },
  severity: { type: "VARCHAR", nullable: false, comment: "flow severity token (see severity list); never NULL ('info')" },
  threat_score: { type: "USMALLINT", nullable: false, comment: "explainable threat score, 0–100" },
  ioc: { type: "BOOLEAN", nullable: false, comment: "true when any threat-feed indicator matched this flow" },
};

/** Engine `category` column tokens (snake_case, matches FlowCategory). */
export const FLOW_CATEGORY_TOKENS = [
  "web",
  "dns",
  "email",
  "file_transfer",
  "remote_access",
  "voip",
  "iot_ot",
  "tunnel_vpn",
  "scan",
  "c2",
  "anomalous",
  "unknown",
  "network_service",
] as const;

/** Engine `severity` column tokens (lowercase; the column is never NULL). */
export const FLOW_SEVERITY_TOKENS = [
  "critical",
  "high",
  "medium",
  "low",
  "info",
] as const;

/** `CREATE TABLE flow (…)` DDL for the in-browser DuckDB, in canonical column order. */
export const FLOW_TABLE_DDL: string = [
  "CREATE TABLE flow (",
  FLOW_COLUMNS.map((name) => {
    const spec = FLOW_COLUMN_TYPES[name];
    return `  ${name} ${spec.type}${spec.nullable ? "" : " NOT NULL"}`;
  }).join(",\n"),
  ");",
].join("\n");
